File: transcription-service/app/services/minio_service.py
# Purpose: Provide reusable service/business logic.

import os
import boto3
import tempfile


# Main flow: Execute core operations and return results.

# Function: get_s3_client - Returns s3 client.
def get_s3_client():
    return boto3.client(
        's3',
        endpoint_url=os.getenv('MINIO_ENDPOINT'),
        aws_access_key_id=os.getenv('MINIO_ACCESS_KEY'),
        aws_secret_access_key=os.getenv('MINIO_SECRET_KEY'),
        region_name='us-east-1',
    )


# Function: download_video - Handles download video.
def download_video(minio_url: str) -> str:
    """Download a video from MinIO to a temp file. Returns the local file path."""
    # Parse the key from the URL
    bucket = os.getenv('MINIO_BUCKET', 'videos')
    key = minio_url.split(f"/{bucket}/", 1)[1]

    s3 = get_s3_client()
    tmp = tempfile.NamedTemporaryFile(delete=False, suffix='.webm')
    s3.download_fileobj(bucket, key, tmp)
    tmp.close()
    return tmp.name

File: upload-service/src/services/minio.service.js
// Purpose: Provide reusable service/business logic.

const { PutObjectCommand, GetObjectCommand } = require('@aws-sdk/client-s3');
const { getSignedUrl } = require('@aws-sdk/s3-request-presigner');
const s3Client = require('../config/minio');

// Main flow: Execute core operations and return results.

// Function: uploadFile - Uploads a file object to MinIO and returns its object URL.
async function uploadFile(key, buffer, contentType) {
  await s3Client.send(new PutObjectCommand({
    Bucket: process.env.MINIO_BUCKET,
    Key: key,
    Body: buffer,
    ContentType: contentType,
  }));

  return `${process.env.MINIO_ENDPOINT}/${process.env.MINIO_BUCKET}/${key}`;
}

// Function: getSignedReadUrl - Returns a temporary signed URL for secure object playback.
async function getSignedReadUrl(key, expiresIn = 3600) {
  const command = new GetObjectCommand({
    Bucket: process.env.MINIO_BUCKET,
    Key: key,
  });

  return getSignedUrl(s3Client, command, { expiresIn });
}

module.exports = { uploadFile, getSignedReadUrl };
